### DISCUSSION TITLE
Zowe SDK - Custom Headers

### DISCUSSION THREAD

**Discussion by Ian-Taggart** [zowe/zowe-cli#1533] [Q&A] [answered]
Hi there,

While using the various z/OSMF methods within "@zowe/zos-jobs-for-zowe-sdk", "@zowe/zos-console-for-zowe-sdk", etc.  - is there any way to add a custom header to those requests making their way out to z/OSMF? 

Reason: Our company has an internal proxy server that has functionality to track requests for various teams based on a custom header and we'd love to use that to track z/OSMF usage through the Zowe SDK. Thank you!

**Comment by FALLAI-Denis** [comment]
Hi,

At the level of the z/OSMF job and spool management APIs, there is a correlator field available to the user.
This could be a way to pass the required information to your proxy.
In this case, the Zowe CLI commands and the Zowe SDK must allow this information to be transmitted.

```
X-IBM-User-Correlator
Specifies the user portion of the job correlator. 
In z/OS, a job correlator can be used to associate each job with a unique 64-character value. 
The correlator provides you with a means to query a job in the system and track it through execution.
A job correlator consists of a 31-byte system-defined portion and a colon character (:), followed by a 32-byte user portion. 
The system-defined portion contains the following values:
8-byte job ID
8-byte MAS name for the system on which the job resides
8-byte sequence value
7-bytes of reserved space.
Following the system value is the colon character (:) separator and the second string: an optional 32-character user-defined value (the user portion). 
This value is 1 — 32 characters in length, where the first character must be uppercase alphabetic (A-Z) or special ($, #, @). 
The remaining characters (up to 31) can be any combination of uppercase alphabetic, numeric (0-9), or special. 
Blank characters are not supported.

If specified, the user portion is combined with the system portion, producing the full job correlator that will be returned in the job-correlator property of the JSON job document. 
If the user portion is not specified, the returned job correlator is the 32-byte system value, ending with the colon (:).

If this header is specified when JES3 is the primary job entry subsystem, an error results and no job is submitted.

For more information on the job correlator, see z/OS JES2 Commands.
```

**Comment by Ian-Taggart** [comment]
Hi Denis, thanks for the reply! Unfortunately, we don't have much control over the naming convention of the proxy headers - this proxy is designed for use with all systems/applications so they have standard naming conventions. Even if the Zowe SDK didn't offer the ability to add custom headers, I've been researching if there's a way to add custom headers for all outgoing connections (regardless of the module) but no luck there either. Not sure if there's an easy answer here (or if it should be made into a request) but I'll keep looking. Thanks again!

**Comment by ATorrise** [comment, marked as answer]
Hello @Ian-Taggart. Currently there is no way to pass a custom header to z/OSMF via the [Zosmf Rest Client](https://github.com/zowe/zowe-cli/blob/master/packages/core/src/rest/ZosmfRestClient.ts). Your question will be turned into an issue requesting an enhancement. If you are still interested in this functionality, the best way to see it come to life is by upvoting the issue and adding any more details of your request [there](https://github.com/zowe/zowe-cli/issues/1611).
